pyomo/contrib/pynumero/algorithms/solvers/pyomo_ext_cyipopt.py

```diff
@@ -0,0 +1,316 @@
+            self._pyomo_nlp.set_primals(primals)
+            ex_inputs = self._ex_io_inputs_from_full_primals(primals)
+            self._ex_io_model.set_inputs(ex_inputs)
+            self._cached_primals = primals.copy()
+
+    def _set_duals_if_necessary(self, duals):
+        if not np.array_equal(duals, self._cached_duals):
+            self._cached_duals.copy_from(duals)
+            self._pyomo_nlp.set_duals(self._cached_duals.get_block(0))
+
+    def _set_obj_factor_if_necessary(self, obj_factor):
+        if obj_factor != self._cached_obj_factor:
+            self._pyomo_nlp.set_obj_factor(obj_factor)
+            self._cached_obj_factor = obj_factor
+
+    def x_init(self):
+        return self._pyomo_nlp.init_primals()
+
+    def x_lb(self):
+        return self._pyomo_nlp.primals_lb()
+    
+    def x_ub(self):
+        return self._pyomo_nlp.primals_ub()
+
+    def g_lb(self):
+        return self._gL.copy()
+
+    def g_ub(self):
+        return self._gU.copy()
+    
+    def objective(self, primals):
+        self._set_primals_if_necessary(primals)
+        return self._pyomo_nlp.evaluate_objective()
+
+    def gradient(self, primals):
+        self._set_primals_if_necessary(primals)
+        return self._pyomo_nlp.evaluate_grad_objective()
+
+    def constraints(self, primals):
+        self._set_primals_if_necessary(primals)
+        pyomo_constraints = self._pyomo_nlp.evaluate_constraints()
+        ex_io_outputs = self._ex_io_model.evaluate_outputs()
+        ex_io_constraints = ex_io_outputs - self._ex_io_outputs_from_full_primals(primals)
+        constraints = BlockVector(2)
+        constraints.set_block(0, pyomo_constraints)
+        constraints.set_block(1, ex_io_constraints)
+        return constraints.flatten()
+
+    def jacobianstructure(self):
+        return self._full_jac_irows, self._full_jac_jcols
+        
+    def jacobian(self, primals):
+        self._set_primals_if_necessary(primals)
+        self._pyomo_nlp.evaluate_jacobian(out=self._jac_pyomo)
+        pyomo_data = self._jac_pyomo.data
+        ex_io_deriv = self._ex_io_model.evaluate_derivatives()
+        # CDL: dense version: ex_io_deriv = self._ex_io_model.evaluate_derivatives().flatten('C')
+        self._full_jac_data[0:len(pyomo_data)] = pyomo_data
+        self._full_jac_data[len(pyomo_data):len(pyomo_data)+len(ex_io_deriv.data)] = ex_io_deriv.data
+        # CDL: dense version: self._full_jac_data[len(pyomo_data):len(pyomo_data)+len(ex_io_deriv)] = ex_io_deriv
+
+        # the -1s for the output variables should still be  here
+        return self._full_jac_data
+
+    def hessianstructure(self):
+        return np.zeros(0), np.zeros(0)
+        #raise NotImplementedError('No Hessians for now')
+
+    def hessian(self, x, y, obj_factor):
+        raise NotImplementedError('No Hessians for now')
+
+    def _ex_io_inputs_from_full_primals(self, primals):
+        return np.compress(self._input_x_mask, primals)
```

Comment:
Does this imply that the primals as they are declared in `self._inputs` are in the same order as the primals in the _pyomo_nlp model?  Is that guaranteed?

Reply:
Another good catch. This will be easy to fix using the _input_columns and _output_columns. I will get to that soon.